### PR TITLE
Create volume on default datastore if default_ds is set for _DEFAULT tenant

### LINKED ISSUE #1044: [DEFAULT tenant] Volume is created on the datastore where vm resides and not on specified default-datastore
VM resides on vsan datastore.

Steps:
1. For default tenant, set sharedVmfs as default datatsore.
2. Created a volume by specifying volume name as short-name - TestVol1
3. Volume gets created on the datastore where vm resides and not on the default-datastore (shared-vmfs).

For user created tenant, specified --default-datastore is picked up during volume creation if volume name is specified using short name.

--------------------------------------

Steps and their output is as follows:

- Verified only _DEFAULT tenant is present with default access rules

```
[root@sc-rdops-vm18-dhcp-32-67:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access ls --name=_DEFAULT
Datastore  Allow_create  Max_volume_size  Total_size
---------  ------------  ---------------  ----------
           True          Unset            Unset

[root@sc-rdops-vm18-dhcp-32-67:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access ls --name=_DEFAULT
Datastore     Allow_create  Max_volume_size  Total_size
------------  ------------  ---------------  ----------
sharedVmfs-0  True          1.00GB           5.00GB
              True          Unset            Unset

```

- Added a datastore rule and set  sharedVmfs-0 as a default-datastore

```
[root@sc-rdops-vm18-dhcp-32-67:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access add --name=_DEFAULT --allow-create --datastore sharedVmfs-0 --volume-ma
xsize=1GB --volume-totalsize=5GB
vm-group access add succeeded

```

- Tried creating a volume using short-name - TestVol1. Volume got created on the datastore where vm resides (vsan datastore) and not on the default-datastore.

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=TestVol1 -o size=1gb
TestVol1
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol1@vsanDatastore
root@sc-rdops-vm02-dhcp-52-237:~#


[root@sc-rdops-vm18-dhcp-32-67:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls
Volume    Datastore      VM-Group  Capacity  Used   Filesystem  Policy          Disk Format  Attached-to  Access      Attach-as               Created By    Created Date
--------  -------------  --------  --------  -----  ----------  --------------  -----------  -----------  ----------  ----------------------  ------------  ------------------------
TestVol1  vsanDatastore  _DEFAULT  1GB       112MB  ext4        [VSAN default]  thin         detached     read-write  independent_persistent  ubuntu-VM0.2  Fri Mar 17 18:44:17 2017

[root@sc-rdops-vm18-dhcp-32-67:~]

```

Logs:
[docker-volume-vsphere.txt](https://github.com/vmware/docker-volume-vsphere/files/851505/docker-volume-vsphere.txt)
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/851506/vmdk_ops.txt)

### PR DESCRIPTION
**Pull request by lipingxue**
Fixed #1044 
This PR includes the fix for:
1.  Datastore used for the first "user created" privilege will be set to "default_ds" for that tenant. 
2.  When creating Docker volume with short name from a VM in  __DEFAULT tenant, use "default_ds" (if set) , not the  vm_datastore

### CODE REVIEW

**Comment by lipingxue** [comment]
I have did the following test manually:
Configuration: VM "photon-6" belongs to _DEFAULT tenant, and this VM is in "datastore1"
1. create a volume with short name, which is created on "datastore1"
```
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description                 Default_datastore  VM_list 
------------------------------------  --------  --------------------------  -----------------  ------- 
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vm-group                

root@photon-KwqUODFXp [ ~ ]# docker volume create --driver=vsphere --name=vol1
vol1
root@photon-KwqUODFXp [ ~ ]#
root@photon-KwqUODFXp [ ~ ]#
root@photon-KwqUODFXp [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere             vol1@datastore1
```

2.  Add access privilege to _DEFAULT tenant with "datastore2", then "datastore2" will become the default_datastore for _DEFAULT tenant
```
/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access add --name=_DEFAULT --datastore=datastore2 --allow-create
vm-group access add succeeded
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4]
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4]
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access ls --name=_DEFAULT
Datastore   Allow_create  Max_volume_size  Total_size 
----------  ------------  ---------------  ---------- 
datastore2  True          Unset            Unset      
            True          Unset            Unset      
 
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4]
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description                 Default_datastore  VM_list 
------------------------------------  --------  --------------------------  -----------------  ------- 
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vm-group  datastore2                 

root@photon-KwqUODFXp [ ~ ]# docker volume create --driver=vsphere --name=vol2
vol2
root@photon-KwqUODFXp [ ~ ]#
root@photon-KwqUODFXp [ ~ ]#
root@photon-KwqUODFXp [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere             vol1@datastore1
vsphere             vol2@datastore2
```

**Comment by lipingxue** [comment]
Per offline discussion with Mark and Sam, I reverted the change in _tenant_access_add.
User need to specify ```--default-datastore``` flag when running ```vm-group access add``` to
set the default datastore for a vm-group.  After setting, user can run ```vm-group ls``` to list the vm-group and the column ```Default_datastore``` will show the default datastore of that vm-group.

**Comment by lipingxue** [comment]
Run the following test manually.
Configuration: VM "photon-6"  is created on "datastore1". This VM belongs _DEFAULT tenant.

step1: Add a privilege for _DEFAULT vm-group with "--default-datastore" flag specified to set the default datastore of _DEFAULT vm-group to "datastore2"
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description               Default_datastore  VM_list  
------------------------------------  --------  ------------------------  -----------------  -------  
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant            

[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access add --name=_DEFAULT --datastore=datastore2 --allow-create --default-datastore
vm-group access add succeeded
[root@localhost:~] 
[root@localhost:~] 
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group access ls --name=_DEFAULT
Datastore   Allow_create  Max_volume_size  Total_size  
----------  ------------  ---------------  ----------  
datastore2  True          Unset            Unset       
            True          Unset            Unset       

```

step2:  check default_datastore of vm-group _DEFAULT
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description               Default_datastore  VM_list  
------------------------------------  --------  ------------------------  -----------------  -------  
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant  datastore2                  

```

step3: create a volume with short name, make sure this volume is created on the default datastore "datastore2" instead of vm datastore "datastore1".
```
root@photon-KwqUODFXp [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
root@photon-KwqUODFXp [ ~ ]# 
root@photon-KwqUODFXp [ ~ ]# docker volume create --driver=vsphere --name=vol1
vol1
root@photon-KwqUODFXp [ ~ ]# 
root@photon-KwqUODFXp [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere             vol1@datastore2

```

**Comment by lipingxue** [comment]
@msterin @shaominchen I have changed this PR per yesterday's offline discussion. Please review it, thanks!

File: esx_service/vmdk_ops.py [hunk 1]
```diff
@@ -806,21 +806,17 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
     if error_info:
         return err(error_info)
-    if tenant_name == auth_data_const.DEFAULT_TENANT:
-        # for DEFAULT tenant, set default_datastore to vm_datastore
+
+    # if default_datastore is not set for tenant,
+    # default_datastore_url will be set to None
+    error_info, default_datastore_url = auth_api.get_default_datastore_url(tenant_name)
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: please update the admin command line message on group creation to something like "vm-group xxx is created. Do not forget to `vm-group vm add` and `vm-group privilege add` to enable better access control" 
(adjust to reflect the correct commands please)

**Comment by lipingxue** [comment]
@msterin @shuklanirdesh82 I have addressed your comments. Please review it.

File: esx_service/cli/vmdkops_admin.py [hunk 2]
```diff
@@ -854,7 +854,8 @@ def tenant_create(args):
     if error_info:
         return operation_fail(error_info.msg)
     else:
-        print("vm-group create succeeded")
+        print("vm-group create succeeded. vm-group '{}' is created. Do not forget to run "
+              "command 'vm-group vm add'  and 'vm-group access add' to enable better access control.".format(args.name))
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: Suggestion:
`"vm-group '{}' is created. Do not forget to run 'vm-group vm add'  and 'vm-group access add' commands to enable  access control.".format(args.name))`

**Pull request merged**